### PR TITLE
Fix a lot of C++ build warnings

### PR DESCRIPTION
**Pull request by jamesr66a**
I went through my build log and did what I thought were reasonable fixes to all the C++ compilation warnings that came up

### CODE REVIEW

**Review by zdevito — APPROVED** [review]

File: aten/src/ATen/SparseTensorImpl.h [hunk 1]
```diff
@@ -88,7 +88,7 @@ struct CAFFE2_API SparseTensorImpl : public TensorImpl {
   // (this could make some of the stored indices out-of-bound and thus unsafe).
   void resize_(int64_t sparse_dim, int64_t dense_dim, IntList size) {
     AT_CHECK(allow_tensor_metadata_change(), "resize_ is not allowed on Tensor created from .data or .detach()");
-    AT_CHECK(sparse_dim + dense_dim == size.size(), "number of dimensions must be sparse_dim (", sparse_dim, ") + dense_dim (", dense_dim, "), but got ", size.size());
+    AT_CHECK(sparse_dim + dense_dim == (int64_t)size.size(), "number of dimensions must be sparse_dim (", sparse_dim, ") + dense_dim (", dense_dim, "), but got ", size.size());
```
Comment (goldsborough) [inline aten/src/ATen/SparseTensorImpl.h]: Please don't use c-style casts

File: aten/src/ATen/core/jit_type.h [hunk 2]
```diff
@@ -308,15 +308,15 @@ struct CAFFE2_API TensorType : public DynamicType {
 
   at::ScalarType scalarType() const { return scalar_type_; }
   at::Device device() const { return device_; }
-  int dim() const { return dim_; }
+  size_t dim() const { return dim_; }
```
Comment (goldsborough) [inline aten/src/ATen/core/jit_type.h]: Are you sure about this? All of ATen uses `int64_t` for sizes and anything integral, so this might cause problems

**Review by goldsborough — APPROVED** [review]
Please fix the c-style casts and check the type of `dim`
